### ISSUE TITLE
RFC: A more concise API for overriding field arguments and consolidating TableSchema and ModelSchema

### ISSUE THREAD

**Issue by sloria** [marshmallow-code/marshmallow-sqlalchemy#240]
### Status quo

Users often need to provide overrides to autogenerated fields.  There are currently two current mechanisms for this.

You can use `field_for`. This gets verbose when you use it for many fields.

```python
class ArtistSchema(ModelSchema):
    class Meta:
        model = models.Artist

    created_at = field_for(models.Artist, "created_at", dump_only=True)
    # ...

```

You can also pass overrides on the model.

```python
class Artist(Model):
    created_at = Column(Timestamp, info={"marshmallow": {"dump_only": True})
```

This awkwardly mixes concerns and can be problematic when you need multiple schemas for the same model.

### Proposed changes

1. Deprecate `ModelSchema` and `TableSchema` in favor of a `SQLAlchemyAutoSchema` that can take either `model` or `table` as an option. Like `ModelSchema` and `TableSchema`, `SQLAlchemyAutoSchema` will autogenerate fields for all columns. Field configuration can be overridden using a new `auto_field` function.

```python
from marshmallow_sqlalchemy import SQLAlchemySchema, auto_field

class ArtistSchema(SQLAlchemyAutoSchema):
    class Meta:
        model = models.Artist
        # OR
        # table = models.Artist.__table__
     
    id = auto_field(dump_only=True)
    created_at = auto_field(dump_only=True)
```

2. Add a `SQLAlchemySchema` class that associates the schema with a model/table so that `auto_field` can be used. Unlike `SQLAlchemyAutoSchema`, fields won't be automatically generated for all columns, so fields must be declared explicitly.

```python
from marshmallow_sqlalchemy import SQLAlchemySchema, auto_field

class ArtistSchema(SQLAlchemySchema):
    class Meta:
        model = models.Artist
        # OR
        # table = models.Artist.__table__

    id = auto_field(dump_only=True)
    created_at = auto_field(dump_only=True)
    name = auto_field()
```

3. Deprecate support for `info={"marshmallow": {...}}`, as it is redundant.

---

h/t @taion for the SQLAlchemySchema/auto_field idea and its implementation (currently outside this repo)

**Comment by sloria** [comment]
@jeanphix @AbdealiJK @deckar01 Thoughts on this API?

**Comment by deckar01** [comment]
> Field configuration can be overridden using a new `auto_field` function.

I really like how concise this API is. Class inheritance is very convenient, so it would be nice to make field inheritance just as convenient. This seems like a syntax that would be useful for extending fields in the core. `fields.override(**kwargs)`?

> `SQLAlchemyAutoSchema`

Is there an advantage to making this a separate class rather than a `Meta` option?

**Comment by sloria** [comment]
> Is there an advantage to making this a separate class rather than a Meta option?

Not a huge advantage. My rationale is that `SQLAlchemySchema` and `SQLAlchemyAutoSchema` behave differently with regards to how they construct the schema at _class declaration time_. This is unlike the Meta options, e.g. `ordered` and `render_module`, which change the behavior of _instances_. This is a low-confidence opinion.

**Comment by AbdealiLoKo** [comment]
I really like this idea.
I have disliked both the current approaches myself:
 - Using `field_for()` is ugly because I need to mention the name of the model for every field. And if I inherit it into another class and change the `Meta.model` is doesn't work.
 - Using `info` is not possible for me, as I have my model centrally defined in a separate module and reuse them in many different schemas

I have a clarification on inheritance:
If I have:
```python
class ArtistSchema(SQLAlchemySchema):
    class Meta:
        model = models.Artist

    id = auto_field(dump_only=True)
    created_at = auto_field(dump_only=True)
    name = auto_field()

class MusicArtistSchema(SQLAlchemySchema):
    class Meta:
        model = models.MusicArtist

    name = auto_field()
    band = auto_field()
```
what is the expected behavior ? I know that 

Regarding Model vs Meta option:
I agree with @sloria , 2 separate classes seems better

**Comment by sloria** [comment]
@AbdealiJK The auto-fields will be generated for subclasses of `SQLAlchemySchema` only. This means that you can use auto_field with mixins like so:

```python
class ArtistMixin:
    id = auto_field(dump_only=True)
    created_at = auto_field(dump_only=True)
    name = auto_field()

class ArtistSchema(ArtistMixin, SQLAlchemySchema):
    class Meta:
        model = models.Artist

class MusicArtistSchema(ArtistMixin, SQLAlchemySchema):
    class Meta:
        model = models.MusicArtist
    band = auto_field()
```

**Comment by sloria** [comment]
@AbdealiJK @deckar01 What do you think about using a type keyword argument?

```python
class ArtistSchema(SQLAlchemySchema, model=models.Artist):
    name = auto_field()
```

This could be a nice typing saver, especially when you have a base schema class with other class Meta options.

With Meta option:

```python
from marshmallow import EXCLUDE
from marshmallow_sqlalchemy import SQLAlchemySchema

class MyBaseModelSchema(SQLAlchemySchema):
    class Meta:
        unknown = EXCLUDE

# -----

class ArtistSchema(MyBaseModelSchema):
    class Meta(MyBaseModelSchema.Meta):
        model = models.ArtistSchema

    name = auto_field()
```

With type kwarg:

```python
from marshmallow import Schema, EXCLUDE
from marshmallow_sqlalchemy import SQLAlchemySchema

class MyBaseModelSchema(SQLAlchemySchema):
    class Meta:
        unknown = EXCLUDE

# -----

class ArtistSchema(MyBaseModelSchema, model=models.Artist):
    name = auto_field()
```

This makes for a very concise API. Some things to consider, though:

1. Is is confusing to use both `class Meta` and type kwargs?
2. A minor implementation weirdness is that `SQLAlchemySchema` subclasses will still need to store the `model` class as a class attribute so that it can later be accessed by the `post_load` method that constructs model instances: https://github.com/marshmallow-code/marshmallow-sqlalchemy/blob/9b85cd80954534fcafe6cbbbf1df83c23bd4ca4f/src/marshmallow_sqlalchemy/schema.py#L184-L185

**Comment by deckar01** [comment]
I think it is interesting to treat field inference as a special type of class inheritance. It does not seem like a very common pattern, so it might not be intuitive to document/learn. Does `Meta` have to subclass its parent's meta? I though we composed the `Meta` classes in the inheritance chain automatically.

<details>
<summary>Decorator syntax tangent</summary>

Another syntax I have played with for registering models is class decorators:

```py
@schema(model=Artist)
class ArtistSchema:
    ...
```

Once I started using a registry that looks up the schema by the model type, I was able to register a schema directly from the model if it didn't need any customization.

```py
@schema()
class Artist:
    ...

artist = load(data, model=Artist)
data = dump(artist) # The schema is automatically inferred from the model instance's class
```

</details>

**Comment by sloria** [comment]
>  Does Meta have to subclass its parent's meta? I though we composed the Meta classes in the inheritance chain automatically.

We do no such magic, so the `Meta` subclassing is necessary.

**Comment by AbdealiLoKo** [comment]
I'm not very in favor of the type kwarg example you mentioned.
It looks neater, but not popular - and the metaclass and other nuances gets a lot of people confused. So, IMO it is confusing

**Comment by taion** [comment]
Would it make sense for a mixin to specify `Meta`, and then for concrete subclasses to override?

Something like:

```python
class WidgetMixin:
    class Meta:
        model = Widget

    name = auto_field()


class WidgetFooSchema(WidgetMixin, Schema):
    foo = auto_field()


class WidgetBarSchema(WidgetMixin, Schema):
    bar = auto_field()
```

This is somewhat contrived though.

**Comment by sloria** [comment]
@taion That would work, but I'm not sure when you'd want to do that.

> I'm not very in favor of the type kwarg

@AbdealiJK Fair enough. The type kwarg is probably too unfamiliar to be worth the saved keystrokes.

**Comment by LukeMarlin** [comment]
Not much to say that wasn't said I think, the proposed API (first post) looks really neat! I'd love to use that instead of the current way as it's:
  - Much more DRY-compliant
  - Names are less confusing: `autofield` is better than `field_for` for me
  - The possibility to go the other way around by explicitly declaring fields but auto-specing them is awesome for any PATCH route where we might want only a few fields

Small note: a decorator seems better than the `Meta`, but I've not enough knowledge about the implications for such a change. And ofc, even if I find it nicer, `Meta` isn't much of an issue  :)

**Comment by caffeinatedMike** [comment]
@sloria This functionality is exactly what I need since I only want to expose maybe half of the columns present in some of our tables and with the current state of the project it's tedious to do so, while maintaining the core objective of being able to update models. Has any work been put into this implementation?

**Comment by sloria** [comment]
We have an implementation of this at work in a closed-source repo. It will take a bit of finagling to get it ready for upstream, but hopefully we'll get to it...soon (sorry I can't give an accurate estimate).

**Comment by sloria** [comment]
This is released in 0.22.0. Please try it out!
